include/dmlc/intrusive_ptr.h

```diff
@@ -0,0 +1,219 @@
+/*!
+ * Copyright (c) by Contributors 2020
+ * \file intrusive_ptr.h
+ * \brief Implementation of Intrusive Ptr.
+ */
+#ifndef DMLC_INTRUSIVE_PTR_H_
+#define DMLC_INTRUSIVE_PTR_H_
+
+#include <atomic>
+#include <cinttypes>
+#include <functional>
+
+namespace dmlc {
+/*!
+ * \brief Helper class for embedding reference counting into client objects.
+ */
+class RefCount {
+  std::atomic<int32_t> count_;
+
+ public:
+  RefCount() noexcept : count_{0} {}
+  std::int32_t Inc() noexcept { return ++count_; }
+  std::int32_t Dec() noexcept { return --count_; }
+  bool IsZero() const { return count_ == 0; }
+  int32_t Count() const { return count_; }
+};
+
+/*!
+ * \brief User defined function for returing embedded reference count.
+ */
+template <typename T> RefCount &IntrusivePtrRefCount(T const *ptr) noexcept;
+/*!
+ * \brief User defined function for destroying object.
+ */
+template <typename T> void IntrusivePtrDestroy(T const *ptr) noexcept;
+
+
+/*!
+ * \brief Implementation of Intrusive Pointer.
+ *
+ * \code
+ *   struct RefCountedData {
+ *     mutable class RefCount ref;
+ *     float data { 0 };
+ *
+ *     friend RefCount &
+ *     IntrusivePtrRefCountRefCount(RefCountedData const *t) noexcept {
+ *       return t->ref;
+ *     }
+ *
+ *     friend void
+ *     IntrusivePtrDestroy(RefCountedData const *t) noexcept {
+ *       delete t;
+ *     }
+ *   };
+ *
+ *   IntrusivePtr p { new RefCountedData };
+ *
+ * \endcode
+ */
+template <typename T> class IntrusivePtr {
+ private:
+  void Incref(T *ptr) {
+    if (ptr) {
+      IntrusivePtrRefCount(ptr).Inc();
+    }
+  }
+  void Decref(T *ptr) {
+    if (ptr) {
+      if (IntrusivePtrRefCount(ptr).Dec() == 0) {
+        IntrusivePtrDestroy(ptr);
+      }
+    }
+  }
+
+ protected:
+  T *ptr_{nullptr};
+
+ public:
+  using element_type = T;  // NOLINT
+  struct Hash {
+    std::size_t operator()(IntrusivePtr<element_type> const &ptr) const noexcept {
+      return std::hash<element_type *>()(ptr.Get());
+    }
+  };
+
+  IntrusivePtr() noexcept = default;
+
+  /*!
+   * \brief Contruct an IntrusivePtr.
+   *
+   * \param p Raw pointer to object
+   * \param add_ref Whether contructing the IntrusivePtr should increase reference count.
+   */
+  explicit IntrusivePtr(element_type *p, bool add_ref = true) : ptr_{p} {
```

Comment:
Do not introduce a raw ptr API, instead create a make_intrusive<T> which creates the intrusive ptr.